metadata-ingestion/src/datahub/ingestion/source/qlik_sense/qlik_api.py

```diff
@@ -102,33 +125,83 @@ def _get_sheet(
             response = websocket_connection.websocket_send_request(method="GetLayout")
             sheet_dict = response[Constant.QLAYOUT]
             sheet = Sheet.parse_obj(sheet_dict[Constant.QMETA])
+            i = 1
             for chart_dict in sheet_dict[Constant.QCHILDLIST][Constant.QITEMS]:
-                sheet.charts.append(Chart.parse_obj(chart_dict[Constant.QINFO]))
+                chart = self._get_chart(
+                    websocket_connection,
+                    chart_dict[Constant.QINFO][Constant.QID],
+                    sheet_id,
+                )
+                if chart:
+                    if not chart.title:
+                        chart.title = f"Object {i}"
+                        i += 1
+                    sheet.charts.append(chart)
+                websocket_connection.handle.pop()
             return sheet
-        except Exception:
-            self._log_http_error(message=f"Unable to fetch sheet with id {sheet_id}")
+        except Exception as e:
+            self._log_http_error(
+                message=f"Unable to fetch sheet with id {sheet_id}. Exception: {e}"
+            )
         return None
 
-    def _get_app_used_datasets(
+    def _add_qri_of_tables(self, tables: List[QlikTable], app_id: str) -> None:
+        table_qri_dict: Dict[str, str] = {}
+        app_qri = quote(f"qri:app:sense://{app_id}", safe="")
+        try:
+            response = self.session.get(
+                f"{self.rest_api_url}/lineage-graphs/nodes/{app_qri}/actions/expand?node={app_qri}&level=TABLE"
+            )
+            response.raise_for_status()
+            for table_node_qri in response.json()[Constant.GRAPH][Constant.NODES]:
+                table_node_qri = quote(table_node_qri, safe="")
+                response = self.session.get(
+                    f"{self.rest_api_url}/lineage-graphs/nodes/{app_qri}/actions/expand?node={table_node_qri}&level=FIELD"
```

Comment:
if the table node is derived from multiple snowflake datasets, does this handle it correctly?

Reply:
We are using lineage graph api just to get the app used table QRI. 
And the actual table details and its dataset/dataconnection it is derived from are been fetched from Websocket json api.
And I just see single data connection details for each tables. Then I guess we don't see situation in which one table derived from multiple snowflake dataset.
Below is response we get for tables used by single qlik app
`{
            "qLayout": {
                "qInfo": {"qId": "LoadModel", "qType": "LoadModel"},
                "tables": [
                    {
                        "dataconnectorName": "Google_BigQuery_harshal-playground-306419",
                        "dataconnectorPrefix": "test_space:",
                        "boxType": "blackbox",
                        "databaseName": "",
                        "ownerName": "",
                        "tableName": "test_table",
                        "tableAlias": "test_table",
                        "loadProperties": {
                            "filterInfo": {"filterClause": "", "filterType": 1}
                        },
                        "key": "Google_BigQuery_harshal-playground-306419:::test_table",
                        "fields": [
                            {
                                "alias": "name",
                                "name": "name",
                                "selected": True,
                                "checked": True,
                                "id": "dsd.test_table.name",
                            }
                        ],
                        "connectionInfo": {
                            "name": "Google_BigQuery_harshal-playground-306419",
                            "displayName": "Google_BigQuery_harshal-playground-306419",
                            "id": "bb5be407-d3d3-4f19-858c-e71d593f09ae",
                            "type": {
                                "provider": "QvOdbcConnectorPackage.exe",
                                "type": "custom",
                                "name": "QvOdbcConnectorPackage",
                                "displayName": "Qlik® ODBC Connector Package",
                                "isStandardConnector": False,
                                "isIframeCompatible": True,
                                "needsConnect": True,
                                "connectDialog": "/customdata/64/QvOdbcConnectorPackage/web/standalone/connect-dialog.html?locale=en-US",
                                "selectDialog": "/customdata/64/QvOdbcConnectorPackage/web/standalone/select-dialog.html?locale=en-US",
                                "selectAddData": "/customdata/64/QvOdbcConnectorPackage/web/standalone/select-adddata.html?locale=en-US",
                                "credentialsDialog": "/customdata/64/QvOdbcConnectorPackage/web/standalone/credentials-dialog.html?locale=en-US",
                                "update": "/customdata/64/QvOdbcConnectorPackage/web/standalone/loadModelUpdate.js",
                                "architecture": {"text": "Common.undefinedbit"},
                                "connectorMain": "QvOdbcConnectorPackage.webroot/connector-main-iframe",
                            },
                            "typeName": "QvOdbcConnectorPackage.exe",
                            "privileges": [
                                "change_owner",
                                "change_space",
                                "delete",
                                "list",
                                "read",
                                "update",
                            ],
                            "sourceConnectorID": "gbq",
                            "dataconnectorPrefix": "test_space:",
                            "isInAppSpace": True,
                            "space": "659d0e41d1b0ecce6eebc9b1",
                            "connectionString": 'CUSTOM CONNECT TO "provider=QvOdbcConnectorPackage.exe;driver=gbq;OAuthMechanism=0;SupportOldClient=true;Catalog_Old=harshal-playground-306419;separateCredentials=false;Catalog=harshal-playground-306419;Min_TLS=1.2;SQLDialect=1;RowsFetchedPerBlock=16384;DefaultStringColumnLength=65535;AllowLargeResults=false;EnableHTAPI=false;HTAPI_MinResultsSize=1000;HTAPI_MinActivationRatio=3;allowNonSelectQueries=false;QueryTimeout=30;Timeout=300;useBulkReader=true;bulkFetchSize=50;rowBatchSize=1;bulkFetchColumnMode=true;maxStringLength=4096;logSQLStatements=false;"',
                            "hasEditableSeparatedCredentials": False,
                            "canDelete": True,
                            "connectorImagePath": "https://iq37k6byr9lgam8.us.qlikcloud.com/customdata/64/QvOdbcConnectorPackage/web/gbq-square.png",
                            "connectorDisplayName": "Google BigQuery",
                            "dbInfo": {},
                        },
                        "id": "dsd.test_table",
                        "tableGroupId": "",
                        "connectorProperties": {
                            "tableQualifiers": [
                                "harshal-playground-306419",
                                "test_dataset",
                            ],
                        },
                        "selectStatement": "SELECT name\nFROM `harshal-playground-306419`.`test_dataset`.`test_table`;",
                        "caching": {"enabled": True, "type": "qvd"},
                    },
                    {
                        "dataconnectorName": "DataFiles",
                        "dataconnectorPrefix": "test_space:",
                        "boxType": "load-file",
                        "databaseName": "IPL_Matches_2022.csv",
                        "ownerName": "TYPE9_CSV",
                        "tableName": "IPL_Matches_2022",
                        "tableAlias": "IPL_Matches_2022",
                        "key": "DataFiles:IPL_Matches_2022.csv:TYPE9_CSV:IPL_Matches_2022",
                        "fields": [
                            {
                                "id": "dsd.IPL_Matches_2022.City",
                                "name": "City",
                                "alias": "City",
                                "selected": True,
                            },
                            {
                                "id": "dsd.IPL_Matches_2022.Date",
                                "name": "Date",
                                "alias": "Date",
                                "selected": True,
                            },
                        ],
                        "connectionInfo": {
                            "type": {"isStorageProvider": False},
                            "id": "87d7bc7e-77d8-40dc-a251-3a35ec107b4e",
                            "name": "DataFiles",
                            "typeName": "qix-datafiles.exe",
                            "sourceConnectorID": "qix-datafiles.exe",
                            "connectionString": 'CUSTOM CONNECT TO "provider=qix-datafiles.exe;path=test_space:datafiles;"',
                            "space": "659d0e41d1b0ecce6eebc9b1",
                            "dataconnectorPrefix": "test_space:",
                            "caching": {"enabled": True, "type": "qvd"},
                        },
                        "id": "dsd.IPL_Matches_2022",
                        "loadSelectStatement": "LOAD [ID] AS [ID],\n\t[City] AS [City],\n\t[Date] AS [Date],\n\t[Season] AS [Season],\n\t[MatchNumber] AS [MatchNumber],\n\t[Team1] AS [Team1],\n\t[Team2] AS [Team2],\n\t[Venue] AS [Venue],\n\t[TossWinner] AS [TossWinner],\n\t[TossDecision] AS [TossDecision],\n\t[SuperOver] AS [SuperOver],\n\t[WinningTeam] AS [WinningTeam],\n\t[WonBy] AS [WonBy],\n\t[Margin] AS [Margin],\n\t[method] AS [method],\n\t[Player_of_Match] AS [Player_of_Match],\n\t[Team1Players] AS [Team1Players],\n\t[Team2Players] AS [Team2Players],\n\t[Umpire1] AS [Umpire1],\n\t[Umpire2] AS [Umpire2]\nFROM [lib://DataFiles/IPL_Matches_2022.csv]\n(txt, codepage is 28591, embedded labels, delimiter is ',', msq);\n",
                        "formatSpec": "(txt, codepage is 28591, embedded labels, delimiter is ',', msq)",
                        "caching": {"enabled": True, "type": "qvd"},
                    },
                ],
                "schemaVersion": 2.1,
            }
        }`